docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+will discuss the properties it requires from any signal handler for
+correctness.
+
+The `KillSwitch` that fires attempts to acquire `terminable`. Because a guest
+fault or timeout has occurred, the guest is actually in
+`lucet_runtime_internals::instance::signals::handle_signal`, and `terminable`
+may be `true` (guest fault), or `false` (handling a timeout). If the guest is
+currently handling a timeout signal, the `KillSwitch` trying to fire will fail
+to acquire `terminable` and exit with an err. However..
+
+#### Timeout while handling a guest fault
+
+In the case that a timeout occurs during a guest fault, the `KillSwitch` may
+acquire `terminable`. POSIX signal handlers are highgly constrained, see `man 7
+signal-safety` for details. The functional constraint imposed on signal
+handlers used with Lucet is that they may not lock on `KillState`'s
+`execution_domain`. As a consequence, a `KillSwitch` may fire during the
+handling of a guest fault - `sigaction` must mask `SIGALRM` so that a signal
+fired before the handler exits is discarded. If the signal behavior is to
+continue without effect, leave termination in place and continue to the guest.
+If, however, the signal handler determines it must return to the host, it
+disables termination on the instance to avoid the case where a timeout occurs
+immediately after swapping to the host context (preventing an erroneous SIGALRM
+in the host context).
+
+### `B -> E` timeout, during guest exit
+
+The `KillSwitch` that fires attempts to acquire `terminable`, but is in a race
+with the teardown in `lucet_context_backstop`. Both functions attempt to swap
+`false` into `terminable`, but only one will see `true` out of it. This acts as
+an indicator for which function may continue, where the other may have to take
+special care as to not leave the instance in a state that would be dangerous to
+signal.
+
+#### Guest acquires `terminable`
+
+The guest exits in the next handful of instructions. The `KillSwitch` that
+failed to acquire the `true` in `terminable` exits with an indication that it
+could not terminate the guest. In this circunstance, we are certain that there
+is a non-conditional and short path out of the guest comprising a return from
+`exit_guest_region` and a context swap back to the host code. Timeouts
+"failing" due to this are only failing because the guest is about to exit, and
+a signal would have no interesting additional benefit.
+
+#### `KillSwitch` acquires `terminable`
+
+In a more unfortunate circumstance, the `KillSwitch` is what observed the
+`true` out of `terminable`. In this case, the guest observed `false` and must
+not proceed, so that whenever an imminent `SIGALRM` from the corresponding
+`KillSwitch` arrives, it will be in a guaranteed-to-be-safe spin loop, or on
+its way there with only signal-safe state.
+
+The `KillSwitch` itself will signal the guest as any other `Domain::Guest`
+interruption.
+
+### `C -> B` timeout
```

Comment:
Rather than handle the state transitions alphabetically, it might be worthwhile to discuss all the straight-forward transitions first and then offer a breakdown of the gnarly transitions to `E`.

Reply:
In thinking this over, I realized my mental model for how someone would read this is kind of like a reference. Rather than reading top to bottom, I expect the reader to be looking for an answer to a specific "how is Lucet correct when ...". I think it's important to be legible top to bottom as well, but I think we'll have a different document that's ordered more amenably to "start reading here" cases. I'm not sure if that's actually the kind of readership this will have, either. Do you have an ordering in mind that satisfies both reading styles?

I'm curious about your thoughts maintaining a different ordering - it feels like it would be confusing to move one of these subsections around if it becomes less straightforward, or move one down if it's simplified in the future. Is that just me overthinking it?

Reply:
My own reading experience of it was educational, "I have no idea how any of this works."  That's quite different from your view of it as a reference.  I support whatever choice you make in the matter.

Reply:
I reordered the sections. If some of them get trickier or simpler, we can just move them - no biggie. :grin: